imap/structure.go

```diff
@@ -3,130 +3,131 @@ package imap
 import (
 	"bytes"
 	"errors"
-	"fmt"
 	"mime"
 	"strings"
 
 	"github.com/ProtonMail/gluon/rfc822"
 )
 
-func Structure(section *rfc822.Section, ext bool) (string, error) {
-	res, err := structure(section, ext)
-	if err != nil {
-		return "", err
+func Structure(section *rfc822.Section) (string, string, error) {
+	bodyBuilder := strings.Builder{}
+	structureBuilder := strings.Builder{}
+
+	writer := dualParListWriter{b1: &bodyBuilder, b2: &structureBuilder}
+
+	c := newParamListWithGroup(&writer)
+	if err := structure(section, &c, &writer); err != nil {
+		return "", "", err
 	}
 
-	return res.String(), nil
+	c.finish(&writer)
+
+	body := bodyBuilder.String()
+	structure := structureBuilder.String()
+
+	return body, structure, nil
 }
 
-func structure(section *rfc822.Section, ext bool) (fmt.Stringer, error) {
+func structure(section *rfc822.Section, fields *paramList, writer *dualParListWriter) error {
 	if len(section.Children()) == 0 {
-		return singlePartStructure(section, ext)
+		return singlePartStructure(section, fields, writer)
 	}
 
-	var fields parList
-
-	children, err := childStructures(section, ext)
-	if err != nil {
-		return nil, err
+	if err := childStructures(section, fields, writer); err != nil {
+		return err
 	}
 
-	fields.add(children)
-
 	header, err := section.ParseHeader()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	_, mimeSubType, mimeParams, err := getMIMEInfo(header)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	fields.add(mimeSubType)
+	fields.addString(writer, mimeSubType)
 
-	if ext {
-		fields.
-			add(mimeParams).
-			add(getDispInfo(header)).
-			add(header.Get("Content-Language")).
-			add(header.Get("Content-Location"))
-	}
+	extWriter := writer.toSingleWriterFrom2nd()
+	fields.addMap(extWriter, mimeParams)
+	getDispInfo(fields, extWriter, header)
```

Comment:
i'd call this `setDispInfo` or `addDispInfo`

Reply:
Done